presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java

```diff
@@ -245,15 +258,29 @@ public void addInput(Page page)
             byte[] rowBytes = output.size() == 0 ? new byte[0] : output.getUnderlyingSlice().byteArray();
             if (shouldReplicate) {
                 for (int i = 0; i < partitionFunction.getPartitionCount(); i++) {
-                    rowBuffer.enqueue(new PrestoSparkRow(i, output.size(), rowBytes));
+                    appendRow(new PrestoSparkRow(i, output.size(), rowBytes));
                 }
                 hasAnyRowBeenReplicated = true;
             }
             else {
                 int partition = getPartition(partitionFunctionArguments, position);
-                rowBuffer.enqueue(new PrestoSparkRow(partition, output.size(), rowBytes));
+                appendRow(new PrestoSparkRow(partition, output.size(), rowBytes));
             }
         }
+        updateMemoryContext();
+    }
+
+    private void appendRow(PrestoSparkRow row)
+    {
+        long rowSize = row.getRetainedSize();
+        if (currentBatchSize + rowSize > BATCH_SIZE) {
+            flush();
+        }
+        if (currentBatch == null) {
+            currentBatch = ImmutableList.builderWithExpectedSize(EXPECTED_ROWS_COUNT_PER_BATCH);
```

Comment:
Does this has to be a `ImutableList.Builder`? -- or a normal `ArrayList` would be good? 

Also maybe consider to use previous list size to initialize new array list capacity -- although shouldn't be a big issue.

Reply:
> Does this has to be a ImutableList.Builder? -- or a normal ArrayList would be good?

We usually prefer immutable collections. But in theory it could also be a regular ArrayList.

> Also maybe consider to use previous list size to initialize new array list capacity -- although shouldn't be a big issue.

Let's keep it like this for now to do not complicate the logic. It can be improved in the future if needed.